### PR TITLE
CronExpression add getTimeBefore function

### PR DESCRIPTION
**Pull request by todaychi**
We can use getTimeBefore to calculate the date which is last time with frequency rule; 
It's opposite from getTimeAfter;

### CODE REVIEW

**Comment by todaychi** [comment]
hi @jhouserizer , can you help me to check if this commit can be merged?

**Comment by zemian** [comment]
Your changes look good, and thank you for the PR @todaychi ! Sorry it took us so long to respond!

However your changes is large, and it requires a Contribution License Agreement (CLA) to be submitted. http://www.quartz-scheduler.org/community/contribute.html

Do you think you can submit that for us so we can merge in your changes?

File: quartz-core/src/main/java/org/quartz/CronExpression.java [hunk 1]
```diff
@@ -321,7 +321,48 @@ public boolean isSatisfiedBy(Date date) {
...
+    }
     
+    public boolean isSatisfied(Date date) {
+        Calendar testDateCal = Calendar.getInstance(getTimeZone());
+        testDateCal.setTime(date);
+        testDateCal.set(Calendar.MILLISECOND, 0);
+
+        Date timeBefore = getTimeBefore(testDateCal.getTime());
+        if (timeBefore == null) {
+            System.out.println("testDateCal " + testDateCal.getTime());
```
Comment (zemian) [inline quartz-core/src/main/java/org/quartz/CronExpression.java]: Please remove System.out.println statement from the main code. If needed, use logger instead.

File: quartz-core/src/main/java/org/quartz/CronExpression.java [hunk 2]
```diff
@@ -321,7 +321,48 @@ public boolean isSatisfiedBy(Date date) {
...
+            return false;
+        }
+        Date originalDate = getTimeAfter(timeBefore);
+        if (originalDate == null) {
+            System.out.println("originalDate " + originalDate);
```
Comment (zemian) [inline quartz-core/src/main/java/org/quartz/CronExpression.java]: Sam here. Remove STDOUT.

File: quartz-core/src/main/java/org/quartz/CronExpression.java [hunk 3]
```diff
@@ -321,7 +321,48 @@ public boolean isSatisfiedBy(Date date) {
...
+        Date originalDate = testDateCal.getTime();
+        
+        testDateCal.add(Calendar.SECOND, 1);
+        
+        Date timeBefore = getTimeBefore(testDateCal.getTime());
+
+        return ((timeBefore != null) && (timeBefore.equals(originalDate)));
+    }
     
+    public boolean isSatisfied(Date date) {
```
Comment (zemian) [inline quartz-core/src/main/java/org/quartz/CronExpression.java]: We should add Javadoc to describe what this new isSatistied() method is for.

**Comment by zemian** [comment]
Please also update PR with latest code master branch so it will get build by the Azure to verify changes.

**Comment by zemian** [comment]
Close due to inactivities

**Pull request closed without merging**